tests/test_compile_kwargs.py

```diff
@@ -0,0 +1,555 @@
+    return model
+
+
+@pytest.mark.parametrize("optimizer", (optimizers_module.SGD, "SGD", "sgd"))
+@pytest.mark.parametrize("n_outputs_", (1, 2))
+def test_optimizer(optimizer, n_outputs_):
+    """Tests compiling of single optimizer with options.
+    Since there can only ever be a single optimizer, there is no
+    ("name", optimizer, "output") option.
+    Only optimizer classes will be compiled with custom options,
+    all others (class names, function names) should pass through
+    untouched.
+    """
+    # Single output
+    X, y = make_classification()
+    y = np.column_stack([y for _ in range(n_outputs_)]).squeeze()
+
+    est = KerasClassifier(
+        model=get_model,
+        num_hidden=20,
+        optimizer=optimizer,
+        optimizer__learning_rate=0.15,
+        optimizer__momentum=0.5,
+        loss="binary_crossentropy",
+    )
+    est.fit(X, y)
+    if isclass(optimizer):
+        np.testing.assert_almost_equal(
+            est.model_.optimizer.momentum.value().numpy(), 0.5, decimal=2
+        )
+        np.testing.assert_almost_equal(
+            est.model_.optimizer.learning_rate.value().numpy(), 0.15, decimal=2
+        )
+    else:
+        # TF actually instantiates an optimizer object even if passed a string
+        # Checking __class__ is not a universally valid check, but it works for the
+        # test optimizer (SGD)
+        assert (
+            est.model_.optimizer.__class__
+            == optimizers_module.get(optimizer).__class__
+        )
+
+
+def test_optimizer_invalid_string():
+    """Tests that a ValueError is raised when an unknown
+    string is passed as an optimizer.
+    """
+
+    X, y = make_classification()
+
+    optimizer = "sgf"  # sgf is not a loss
+
+    est = KerasClassifier(
+        model=get_model,
+        num_hidden=20,
+        optimizer=optimizer,
+        loss="binary_crossentropy",
+    )
+    with pytest.raises(ValueError, match="Unknown optimizer"):
+        est.fit(X, y)
+
+
+def test_compiling_of_routed_parameters():
+    """Tests that routed parameters
+    can themselves be compiled.
+    """
+
+    X, y = make_classification()
+
+    class Foo:
+        got = dict()
+
+        def __init__(self, foo_kwarg):
+            self.foo_kwarg = foo_kwarg
+
+    class MyLoss:
+        def __init__(self, param1):
+            self.param1 = param1
+            self.__name__ = str(id(self))
+
+        def __call__(self, y_true, y_pred):
+            return losses_module.binary_crossentropy(y_true, y_pred)
+
+    est = KerasClassifier(
+        model=get_model,
+        loss=MyLoss,
+        loss__param1=[Foo, Foo],
+        loss__param1__foo_kwarg=1,
+        loss__param1__0__foo_kwarg=2,
+    )
+    est.fit(X, y)
+    assert est.model_.loss.param1[0].foo_kwarg == 2
+    assert est.model_.loss.param1[1].foo_kwarg == 1
+
+
+@pytest.mark.parametrize(
+    "loss", (losses_module.BinaryCrossentropy, "BinaryCrossentropy",),
+)
+@pytest.mark.parametrize("n_outputs_", (1, 2))
```

Comment:
Why is it necessary to have `n_outputs_` in this test?

Reply:
I'm seeing `n_outputs_` in a lot of tests. Which tests is it necessary in?

Reply:
All of them. Keras changes how it names and arranges losses depending on the number of outputs.

For example, for a single output `Model(...).loss` is a list with a single item (the overall loss is the same loss as the single output's loss). For multi-output models, `Model(...).loss` is a list where the first item is still the overall loss, but there's also an additional entry for each output.

It also changes the behavior when naming losses. `loss=["binary_crossentropy"]` for a single model results in `model_.loss[0].name == "binary_crossentropy"` while `loss=["binary_crossentropy", "binary_crossentropy"]` with 2 outputs yields `model_.loss[0].name == "out1_binary_crossentropy"` IIRC.